### PR TITLE
Use mutating matrix multiplication in eigs (retake)

### PR DESCRIPTION
**Pull request by p-zubieta**
This is basically JuliaLang/julia#7907 with @ViralBShah comments taken into account. I decided to do it here since that seemed almost ready to me.

There are similar things that can be done to reduce allocations in various cases but that would require JuliaLang/LinearAlgebra.jl#216 to be solved first and can be done later.

CC @andreasnoack

### CODE REVIEW

File: base/linalg/arnoldi.jl [hunk 1]
```diff
@@ -237,7 +237,11 @@ function _eigs(A, B;
     end
 
     # Refer to ex-*.doc files in ARPACK/DOCUMENTS for calling sequence
-    matvecA(x) = A * x
+    if method_exists(Base.A_mul_B!, (Vector{T},typeof(A),Vector{T}))
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: Probably much cleaner to just replace uses of `matvecA` with `A_mul_B!` here. Using `method_exists` is going to be slow and difficult to predict.

**Comment by pabloferz** [comment]
Ok, it seems that neither `sub` nor `slice` would work in general, as pointed out in #7907.

Should we use `pointer_to_array` instead?

**Comment by pabloferz** [comment]
It seems that the problem is that `\(A::Sparse, x)` won't work for now if `x` is, for example, an `StridedVecOrMat` instead of a `VecOrMat`.

Using `pointer_to_array` is probably not a good idea, and it would probably be better to wait until all the work being done by @timholy and @mbauman settles down to restructure some of linear algebra functions. Probably in this case it would suffice to define '\(A::Sparse, x::FastSubArray)'.

File: base/linalg/arnoldi.jl (hunk 1 above)
Comment (pabloferz) [inline base/linalg/arnoldi.jl]: I change this but the problem now is that `A_mul_B!` won't work for custom types with `*` defined but without indexing (see the last failure).

The problem can be ultimately tracked down to `generic_matvecmul` which is defined in terms of indexing operations, but for a type that is an array container with only custom `*` overloaded it will fail.

**Comment by andreasnoack** [comment]
@pabloferz Thanks for working on this. I agree with @tkelman that it would be better to completely remove `matvecA` and just use `A_mul_B!`.

You should also use `SubArray`s here instead of `pointer_to_array`. The latter should generally be avoided. If this doesn't work, then we'll need to look into why it doesn't work.

What is the problem with `\(A::Sparse, xSubArray)`?

**Comment by pabloferz** [comment]
@andreasnoack I agree that it would be better just using `A_mul_B!` and `SubArray`s here. Nevertheless each posses its own problem.

On the `A_mul_B!` side the `arnoldi.jl` test will fail given that `A_mul_B!` falls back to `generic_matvecmul` in this case, but the last one is defined in terms of indexing and the test in `arnoldi.jl` define a custom wrapper around an array with no indexing defined but just `*` overloaded. 

An strategy to solve this, would be to write a `generic_matvecmul` that works as `copy!(y, A*X)` when there is no indexing defined for the custom type (but, there are probably other ways to work around this)

On the other hand, `eigs` also uses `solveSI` which is basically `\' or `A_ldiv_B!`. The problem is actually with the later, for example `A_ldiv_B!(::Base.SparseMatrix.CHOLMOD.Factor{Float64}, ::Array{Float64,1})` won't work.

**Comment by pabloferz** [comment]
More precisely `\(::Base.SparseMatrix.CHOLMOD.Factor{Float64}, ::xSubArray)` ends up calling <a href="https://github.com/JuliaLang/julia/blob/2ac304dfba75fad148d4070ef4f8a2e400c305bb/base/linalg/factorization.jl#L23">this method</a> which relies on `A_ldiv_B!` which in turn is not defined for some types.

**Comment by pabloferz** [comment]
The `solveSI` case can be solved by defining

```julia
(\)(L::Factor, b::StridedVector) = Vector(L\convert(Dense{eltype(L)}, b))
function Base.convert{T<:VTypes}(::Type{Dense{T}}, A::StridedVecOrMat)
    d = allocate_dense(size(A, 1), size(A, 2), stride(A, 2), T)
    s = unsafe_load(d.p)
    for i in eachindex(A)
        unsafe_store!(s.x, A[i], i)
    end
    d
end
function Base.convert(::Type{Dense}, A::StridedVecOrMat)
    T = promote_type(eltype(A), Float64)
    return convert(Dense{T}, A)
end
```

those methods only work for `VecOrMat` and `Vector` instead.

**Comment by andreasnoack** [comment]
> On the `A_mul_B!` side the `arnoldi.jl` test will fail given that `A_mul_B!` falls back to `generic_matvecmul` in this case, but the last one is defined in terms of indexing and the test in `arnoldi.jl` define a custom wrapper around an array with no indexing defined but just `*` overloaded.

I'm not sure I follow. `*` just allocates a return array and calls `A_mul_B!`. Why is `A_mul_B!` falling back to `generic_matvecmul!`? See e.g.
```jl
julia> @which A_mul_B!(ones(10), speye(10), ones(10))
A_mul_B!(C::Union{Base.ReshapedArray{T<:Any,1,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},Base.ReshapedArray{T<:Any,2,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},DenseArray{T<:Any,1},DenseArray{T<:Any,2},SubArray{T<:Any,1,A<:Union{Base.ReshapedArray{T<:Any,N<:Any,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},DenseArray},I<:Tuple{Vararg{Union{Base.AbstractCartesianIndex,Base.NoSlice,Colon,Int64,Range{Int64}},N<:Any}},L<:Any},SubArray{T<:Any,2,A<:Union{Base.ReshapedArray{T<:Any,N<:Any,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},DenseArray},I<:Tuple{Vararg{Union{Base.AbstractCartesianIndex,Base.NoSlice,Colon,Int64,Range{Int64}},N<:Any}},L<:Any}}, A::SparseMatrixCSC, B::Union{Base.ReshapedArray{T<:Any,1,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},Base.ReshapedArray{T<:Any,2,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},DenseArray{T<:Any,1},DenseArray{T<:Any,2},SubArray{T<:Any,1,A<:Union{Base.ReshapedArray{T<:Any,N<:Any,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},DenseArray},I<:Tuple{Vararg{Union{Base.AbstractCartesianIndex,Base.NoSlice,Colon,Int64,Range{Int64}},N<:Any}},L<:Any},SubArray{T<:Any,2,A<:Union{Base.ReshapedArray{T<:Any,N<:Any,A<:DenseArray,MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64},N<:Any}}},DenseArray},I<:Tuple{Vararg{Union{Base.AbstractCartesianIndex,Base.NoSlice,Colon,Int64,Range{Int64}},N<:Any}},L<:Any}}) at sparse/linalg.jl:101
```

You are right that `A_ldiv_B!` is a problem. The issue is that SuiteSparse doesn't really support in-place solves and we are using SuiteSparse for our sparse solves. Here, we'll probably have to stick with `\` for a now. However, it shouldn't hold back the changes to `*`.

**Comment by pabloferz** [comment]
Thinking a bit more about it, this shows that a restructuring of `eigs` and the `LinAlg` and sparse methods APIs is needed.

A well place to start seems to be #11325

**Comment by pabloferz** [comment]
@andreasnoack The problem is, for example, <a href="https://github.com/JuliaLang/julia/blob/91c0db76c8b5a97b8657d6384d3f595d04de75a6/test/linalg/arnoldi.jl#L130">in this test in base</a>, were you have something like

```julia
type CPM{T<:Base.LinAlg.BlasFloat}<:AbstractMatrix{T} # completely positive map
    kraus::Array{T,3} # kraus operator representation
end
function *{T<:Base.LinAlg.BlasFloat}(Phi::CPM{T},rho::Vector{T})
    rho=reshape(rho,(size(Phi.kraus,3),size(Phi.kraus,3)))
    rho2=zeros(T,(size(Phi.kraus,1),size(Phi.kraus,1)))
    for s=1:size(Phi.kraus,2)
        As=slice(Phi.kraus,:,s,:)
        rho2+=As*rho*As'
    end
    return reshape(rho2,(size(Phi.kraus,1)^2,))
end
```

`eigs` will fail here, as there is no `A_mul_B!` defined for this type, so it will fallback to `generic_matvecmul!`

**Comment by pabloferz** [comment]
I agree that `A_ldiv_B!` should be left for later leaving `\` for now.

**Comment by andreasnoack** [comment]
The `CPM` example should just be adjusted to define `A_mul_B!` instead of `*`, i.e. `rho2` should be an argument to the function.

**Comment by pabloferz** [comment]
Ok, I will chage it. Anyway the definitions in the <a href="https://github.com/JuliaLang/julia/pull/16294#issuecomment-220002369">comment above</a> are still needed, given that the same `SubArray` will be used by `matmulA` and `solveSI`.

I can send another PR to tackle that.

**Comment by andreasnoack** [comment]
> Anyway the definitions in the comment above are still needed, given that the same `SubArray` will be used by `matmulA` and `solveSI`.

I don't understand why you'd have to change anything for `solveSI`. You might be right but could you explain it with more details.

**Comment by pabloferz** [comment]
Ok, sorry for not being clearer before. Let me try again.

In the `aupd_wrapper`, depending on the `mode` and `ido` values a method is selected, in some cases it is `matvecA` in other cases it is `solveSI`. Irrespectively of the case selected, this methods take the same input, which now is going to be a `SubArray`.

When `solveSI` is selected this calls `\`. In the particular case of `\(F::Base.SparseMatrix.CHOLMOD.Factor, ::SubArray)`, which is not specifically defined for this signature, `\` will fall back to `\(F::Factorization, B::AbstractArray)`. The later is defined in term of `A_ldiv_B!` and this will fail (as you noted above, given how SuiteSparse works).

A way around this is to have `\(F::Base.SparseMatrix.CHOLMOD.Factor, ::SubArray)` as I was proposing before.

**Comment by andreasnoack** [comment]
Thanks. Now I understand and I think that your proposed solution in https://github.com/JuliaLang/julia/pull/16294#issuecomment-220002369 is the right way to fix this.

**Comment by pabloferz** [comment]
I had to add an `A_mul_B!` for `SVDOperator` or the `svds` tests would fail. It seems to be working now, except for one of the AppVeyor builds (I don't know if the error is related to de PR or not).

**Comment by pabloferz** [comment]
Ok. Seems like all test passed now. Is everything all right. Should I change something else?

**Comment by tkelman** [comment]
should `matvecA` be renamed `matvecA!` if it's mutating?

**Comment by pabloferz** [comment]
Done. But, one of the AV server builds became unresponsive. Since I only changed `matvecA` to `matvecA!` and it passed before this I guess it shouldn't be a problem to merge this.

**Comment by pabloferz** [comment]
Ok, I think this is ready, but I'd be happy to change anything if this can be further improved.

File: base/linalg/arnoldi.jl [hunk 2]
```diff
@@ -296,6 +296,12 @@ end
 
 ## v = [ left_singular_vector; right_singular_vector ]
 *{T,S}(s::SVDOperator{T,S}, v::Vector{T}) = [s.X * v[s.m+1:end]; s.X' * v[1:s.m]]
```
Comment (ViralBShah) [inline base/linalg/arnoldi.jl]: Is this method needed any longer? If so, should we rewrite `*` here in terms of `A_mul_B!`, since that should use fewer allocations.

**Comment by ViralBShah** [comment]
Apart from my comment above, I think this appears great, and ready to go.

**Comment by ViralBShah** [comment]
If we have an example of good performance improvement here, it is worth noting in NEWS as well.

File: base/linalg/arnoldi.jl (hunk 2 above)
Comment (pabloferz) [inline base/linalg/arnoldi.jl]: I don't think this is needed anywhere else. It was just used by the `eigs` function so no harm should arise from removing it.

File: base/linalg/arnoldi.jl [hunk 3]
```diff
@@ -237,7 +237,7 @@ function _eigs(A, B;
     end
 
     # Refer to ex-*.doc files in ARPACK/DOCUMENTS for calling sequence
-    matvecA(x) = A * x
+    matvecA!(y, x) = A_mul_B!(y, A, x)
     if !isgeneral           # Standard problem
         matvecB = x -> x
```
Comment (andreasnoack) [inline base/linalg/arnoldi.jl]: Can't `matvecB` be changed to in-place as well, i.e. `copy!` here and `A_mul_B!` below?

Comment (pabloferz) [inline base/linalg/arnoldi.jl]: I tried that but <a href="https://github.com/JuliaLang/julia/pull/16294/files#diff-503c2ea34331f7c62b228dd425a7539eR90">here</a> we have `y[:] = solveSI(matvecB(x))`, and that seemed to slow things down in some cases.

I would have guessed that `copy!(y,x)` would be effectively result in the identity when `y === x` (maybe it is if `memmove` takes care of that, but I saw a performance hit anyway).

**Comment by pabloferz** [comment]
> If we have an example of good performance improvement here, it is worth noting in NEWS as well.

Does one of the `@nanosoldier` tests cover this?

File: base/linalg/arnoldi.jl (hunk 2 above)
Comment (ViralBShah) [inline base/linalg/arnoldi.jl]: Could you remove it? I think we can merge this.

**Comment by ViralBShah** [comment]
I think an improvement in memory utilization is also great. Let's merge once we delete the `*` method discussed above.

**Comment by tkelman** [comment]
should be squashed upon merge if the first commit would fail tests

**Comment by ViralBShah** [comment]
Test failure looks not relevant.

**Comment by yuyichao** [comment]
The failure is an oom. I'll try to come up with a better restarting threshold soon...
